libs/MeshKernel/src/Polygon.cpp

```diff
@@ -254,52 +254,97 @@ std::vector<double> meshkernel::Polygon::EdgeLengths() const
     return edgeLengths;
 }
 
+// Refines the section from node specified by index to the next node. The first node is added, the next node is not
+void RefineSegment(std::vector<meshkernel::Point>& refinedPolygon,
+                   const std::vector<meshkernel::Point>::const_iterator& nodeIterator,
+                   const double refinementDistance,
+                   const meshkernel::Projection projection)
+{
+    // Line segment starting point.
+    const auto& n0 = *nodeIterator;
+    const auto& n1 = *std::next(nodeIterator);
+
+    refinedPolygon.push_back(n0);
+
+    const double segmentLength = ComputeDistance(n0, n1, projection);
+
+    int n = std::lround(segmentLength / refinementDistance);
+    const double refinedLength = n * refinementDistance;
+    if (refinedLength > segmentLength ||
+        meshkernel::IsEqual(refinedLength, segmentLength, meshkernel::constants::geometric::refinementTolerance))
+        --n;
+
+    // Refined segment step size.
+    const meshkernel::Point delta = (n1 - n0) * refinementDistance / segmentLength;
+    for (auto i = 1; i <= n; ++i)
+    {
+        refinedPolygon.push_back(n0 + i * delta);
+    }
+}
+
 std::vector<meshkernel::Point> meshkernel::Polygon::Refine(const size_t startIndex, const size_t endIndex, const double refinementDistance) const
 {
+    if (startIndex == endIndex)
+    {
+        return m_nodes;
+    }
 
-    if (startIndex > endIndex || endIndex >= m_nodes.size())
+    if (startIndex >= m_nodes.size() || endIndex >= m_nodes.size())
```

Comment:
You can also use [Range checks](https://github.com/Deltares/MeshKernel/blob/master/libs/MeshKernel/include/MeshKernel/RangeCheck.hpp).

Reply:
nice! done.